components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java

```diff
@@ -1505,6 +1510,93 @@ public static SpOAuth2ExpiryTimeConfiguration getSpTokenExpiryTimeConfig(String
         return spTokenTimeObject;
     }
 
+    /**
+     * Retrieve audience configured for the particular service provider.
+     *
+     * @param clientId
+     * @param tenantDomain
+     * @return
+     */
+    public static List<String> getOIDCAudience(String clientId, String tenantDomain) {
+
+        List<String> oidcAudiences = getDefinedCustomOIDCAudiences(clientId, tenantDomain);
+        // Need to add client_id as an audience value according to the spec.
+        if (!oidcAudiences.contains(clientId)) {
+            oidcAudiences.add(0, clientId);
+        } else {
+            Collections.swap(oidcAudiences, oidcAudiences.indexOf(clientId), 0);
+        }
+        return oidcAudiences;
+    }
+
+    private static List<String> getDefinedCustomOIDCAudiences(String clientId, String tenantDomain) {
+
+        List<String> audiences = new ArrayList<>();
+
+        // Priority should be given to service provider specific audiences over globally configured ones.
+        if (OAuth2ServiceComponentHolder.isAudienceEnabled()) {
+            OIDCAudienceCacheKey cacheKey = new OIDCAudienceCacheKey(clientId + "@" + tenantDomain);
+            CacheEntry result = OIDCAudienceCache.getInstance().getValueFromCache(cacheKey);
+
+            // Check cache hit, do the type check.
+            if (result != null && result instanceof OIDCAudienceCacheEntry) {
+                audiences = ((OIDCAudienceCacheEntry) result).getAudiences();
+                return audiences;
+            } else {
+                OAuthAppDAO oAuthAppDAO = new OAuthAppDAO();
+                try {
+                    audiences = oAuthAppDAO.getOIDCAudiences(tenantDomain, clientId);
+                    if (CollectionUtils.isNotEmpty(audiences)) {
+                        OIDCAudienceCacheEntry cacheEntry = new OIDCAudienceCacheEntry();
+                        cacheEntry.setAudiences(audiences);
+                        OIDCAudienceCache.getInstance().addToCache(cacheKey, cacheEntry);
```

Comment:
Why do we need a separate cache for OIDC Audience?
Can't we use the OAuthAppDO to fetch the OIDC audience?

Reply:
Yes, +1 for that. Check whether there is any specific reason to do this.